File: Day17_Triangle.cpp
#include <bits/stdc++.h>
using namespace std;

int minimumTotal(vector<vector<int>> &triangle)
{
    int sz = triangle.size();
    int n = triangle[sz - 1].size();
    int dp[sz][n];
    for (int i = sz - 1; i >= 0; i--)
    {
        int n1 = triangle[i].size();
        for (int j = 0; j < n1; j++)
        {
            if (i == sz - 1)
            {
                dp[i][j] = triangle[i][j];
                continue;
            }
            dp[i][j] = triangle[i][j] + min(dp[i + 1][j], dp[i + 1][j + 1]);
        }
    }

    return dp[0][0];
}

int main()
{
    vector<vector<int>> v = {{2}, {3, 4}, {6, 5, 7}, {4, 1, 8, 3}};
    cout << minimumTotal(v) << endl;
    return 0;
}

File: Day17_Triangle.py
def minimumTotal(self, triangle: list[list[int]]) -> int:
    sz = len(triangle)
    n = len(triangle[sz-1])
    dp = [0 * n] * sz
    for i in range(sz-1, -1, -1):
        n1 = len(triangle[i])
        for j in range(n1):
            if(i == (sz-1)):
                dp[i][j] = triangle[i][j]
                continue
            dp[i][j] = triangle[i][j] + min(dp[i+1][j], dp[i+1][j+1])

    return dp[0][0]